tools/server/webui/src/utils/app.context.tsx

```diff
@@ -244,37 +255,147 @@ export const AppContextProvider = ({
         body: JSON.stringify(params),
         signal: abortController.signal,
       });
+
       if (fetchResponse.status !== 200) {
         const body = await fetchResponse.json();
         throw new Error(body?.error?.message || 'Unknown error');
       }
-      const chunks = getSSEStreamAsync(fetchResponse);
-      for await (const chunk of chunks) {
-        // const stop = chunk.stop;
-        if (chunk.error) {
-          throw new Error(chunk.error?.message || 'Unknown error');
+
+      // Tool calls results we will process later
+      const pendingMessages: PendingMessage[] = [];
+      let lastMsgId = pendingMsg.id;
+      let shouldContinueChain = false;
+
+      if (params.stream) {
```

Comment:
I know we don't yet support stream with tool_calls, but we will have it soon and I think this whole `if..else` may become redundant in the future

What I'm thinking is to abstract out this, maybe into a func `createChatCompletionStream`, and from outside perspective the function will always return streamed response.

Inside, we if we see `tool_calls`, we disable `stream`while still converts the non-stream response to streamed. In the near future, when we have streamed tool_calls support, we can simply delete that code branch in `createChatCompletionStream`

Reply:
This is right, though ... I'm not really confident I can do it well enough tbf :/, at least not in this PR as I feel like it will grow even more.

Reply:
I can help doing this if you can take care of the other concern about weak typing

Btw, even though we don't yet have mcp, I think having a placeholder class for it (just to demo how your code is designed to handle such case) can be a good idea